src/psij/executors/rest_service.py

```diff
@@ -0,0 +1,163 @@
+"""This module contains the RestService :class:`~psij.JobExecutor`."""
+
+import json
+import requests
+import websocket
+import threading
+
+from urllib.parse import urlparse, urlunparse
+
+from typing import Optional, List, Dict
+
+from psij import (
+    Job,
+    JobExecutorConfig,
+    JobState,
+    JobStatus,
+    JobExecutor,
+    Export,
+)
+
+
+class RestServiceJobExecutor(JobExecutor):
+    """A :class:`~psij.JobExecutor` for a PSI/J Rest service endpoint.
+
+    This executor forwards all requests to a Rest service endpoint which is then
+    executing the respective request on the target resource.
+    """
+
+    _state_map = {'NEW': JobState.NEW,
+                  'QUEUED': JobState.QUEUED,
+                  'ACTIVE': JobState.ACTIVE,
+                  'COMPLETED': JobState.COMPLETED,
+                  'FAILED': JobState.FAILED,
+                  'CANCELED': JobState.CANCELED}
+    _final = [JobState.COMPLETED, JobState.FAILED, JobState.CANCELED]
+
+    def __init__(
+        self, url: Optional[str] = None, config: Optional[JobExecutorConfig] = None
+    ) -> None:
+        """
+        Initializes a `RestServiceJobExecutor`.
+
+        :param url: address at which to contact the remote service.
+                    Supported schemas: `http://` and `https://`
+        :param config: The `RestServiceJobExecutor` does not have any
+                    configuration options.
+        """
+        if not config:
+            config = JobExecutorConfig()
+
+        if not url:
+            url = 'http://localhost:8000'
+
+        url_obj = urlparse(url)
+
+        if url_obj.scheme not in ['http', 'https']:
+            raise ValueError('expected `http://` or `https://` as url schema')
+
+        if url_obj.path and url_obj.path != '/':
+            name = url_obj.path.strip('/')
+        else:
+            name = 'local'
+
+        url = urlunparse(url_obj._replace(path=''))
+
+        super().__init__(url=str(url), config=config)
+
+        self._jobs: Dict[str, Job] = dict()
+        self._idmap: Dict[str, str] = dict()
+        self._serialize = Export()
+
+        # we can only update the idmap when the `submit` request returns, but
+        # an state notification may happen before that.  We use a lock to ensure
+        # that state updates are delayed until after `submit` completed.
+        self._lock = threading.Lock()
+
+        # connect to service and register this client instance
+        self._cid = str(requests.get('%s/executor/%s' % (url, name)).json())
+
+        # create a daemong thread for websocket state notifications
+        t = threading.Thread(target=self._state_listener)
+        t.daemon = True
+        t.start()
+
+    def __del__(self) -> None:
+        """Stop websocket thread upon destruction."""
+        ...
```

Comment:
should be ``pass`` I think, or else nothing (it's not a syntax error as long as there's a docstring). Not that it's an error as is though

Reply:
But since this method isn't used, should just be removed?

Reply:
Oh, right, that's a leftover from a different approach.  Thanks, removed it now.